src/compiler/program.rs

```diff
@@ -0,0 +1,169 @@
+use std::rc::Rc;
+
+use crate::{
+    create_source_file, for_each, get_sys, normalize_path, to_path as to_path_helper, CompilerHost,
+    CreateProgramOptions, Diagnostic, ModuleResolutionHost, Path, Program, SourceFile,
+    StructureIsReused, System,
+};
+
+fn create_compiler_host() -> impl CompilerHost {
+    create_compiler_host_worker()
+}
+
+struct CompilerHostConcrete {
+    system: &'static dyn System,
+}
+
+impl ModuleResolutionHost for CompilerHostConcrete {
+    fn read_file(&self, file_name: &str) -> Option<String> {
+        self.system.read_file(file_name)
+    }
+}
+
+impl CompilerHost for CompilerHostConcrete {
+    fn get_source_file(&self, file_name: &str) -> Option<SourceFile> {
+        let text = self.read_file(file_name);
+        text.map(|text| create_source_file(file_name, &text))
+    }
+
+    fn get_canonical_file_name(&self, file_name: &str) -> String {
+        file_name.to_string()
+    }
+
+    fn get_current_directory(&self) -> String {
+        self.system.get_current_directory()
+    }
+}
+
+fn create_compiler_host_worker() -> impl CompilerHost {
+    CompilerHostConcrete { system: get_sys() }
+}
+
+struct ProgramConcrete {
+    files: Vec<Rc<SourceFile>>,
+}
+
+impl ProgramConcrete {
+    pub fn new(files: Vec<Rc<SourceFile>>) -> Self {
+        ProgramConcrete { files }
+    }
+
+    fn get_diagnostics_helper(
+        &self,
+        get_diagnostics: fn(&ProgramConcrete, &SourceFile) -> Vec<Box<dyn Diagnostic>>,
+    ) -> Vec<Box<dyn Diagnostic>> {
+        self.get_source_files()
+            .iter()
+            .flat_map(|source_file| get_diagnostics(self, source_file))
+            .collect()
+    }
+
+    fn get_semantic_diagnostics_for_file(
+        &self,
+        _source_file: &SourceFile,
+    ) -> Vec<Box<dyn Diagnostic>> {
+        vec![]
+    }
+}
+
+impl Program for ProgramConcrete {
+    fn get_semantic_diagnostics(&self) -> Vec<Box<dyn Diagnostic>> {
+        self.get_diagnostics_helper(ProgramConcrete::get_semantic_diagnostics_for_file)
+    }
+
+    fn get_source_files(&self) -> &[Rc<SourceFile>] {
+        &self.files
+    }
+}
+
+struct CreateProgramHelperContext<'a> {
+    processing_other_files: &'a mut Vec<Rc<SourceFile>>,
+    host: &'a dyn CompilerHost,
+    current_directory: &'a str,
+}
+
+pub fn create_program(root_names_or_options: CreateProgramOptions) -> impl Program {
+    let CreateProgramOptions { root_names } = root_names_or_options;
+
+    let mut processing_other_files: Option<Vec<Rc<SourceFile>>> = None;
+    let mut files: Vec<Rc<SourceFile>> = vec![];
+
+    let host = create_compiler_host();
+
+    let current_directory = host.get_current_directory();
+
+    let structure_is_reused = StructureIsReused::Not;
+    if !matches!(structure_is_reused, StructureIsReused::Completely) {
```

Comment:
I'd use `==` here too rather than `matches!`.

Reply:
Ok updated to derive `Eq, PartialEq` for `StructureIsReused` and use `==` here